src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -1139,7 +1141,7 @@ private async Task WriteNavigationLinksAsync(SelectExpandNode selectExpandNode,
             Contract.Assert(selectExpandNode != null);
             Contract.Assert(resourceContext != null);
 
-            if (selectExpandNode.SelectedNavigationProperties == null)
+            if (selectExpandNode.SelectedNavigationProperties == null || resourceContext.IsPostRequest)
```

Comment:
I noticed something weird in this logic.
For some background, all navigation properties that are not "expanded" are in the `SelectedNavigationProperties`.
So for each, we call:
```csharp
    await writer.WriteStartAsync(nestedResourceInfo);
    await writer.WriteEndAsync();
```
Weirdly, the above logic doesn't write anything in the output.
This logic has issues when serializing deep insert resources. In this case, we haven't expanded anything so we will call the above methods in the `WriteNavigationLinksAsync` logic then call the methods below in `WriteNavigationPropertiesForDeepInsertAsync` method.
```csharp
    await writer.WriteStartAsync(nestedResourceInfo);
    await WriteComplexAndExpandedNavigationPropertyAsync(navigationProperty.Key, null, resourceContext, writer);
    await writer.WriteEndAsync();
```
Calling `WriteStartAsync(nestedResourceInfo);` twice causes the `DuplicatePropertyNameChecker` to throw an exception thinking we are writing 2 nested properties with the same name.

This is why for deep insert we avoid calling `WriteStartAsync(nestedResourceInfo);` in the the `WriteNavigationLinksAsync` method since we will call it again in the `WriteNavigationPropertiesForDeepInsertAsync` method.

Reply:
@KenitoInc Based on your explanation above, that logic feels a bit brittle

Reply:
> Weirdly, the above logic doesn't write anything in the output.

There is some metadata which is only emitted when serializing in full-metadata mode. I think that could be one of them.